tools/autograd/gen_variable_type.py

```diff
@@ -537,16 +574,82 @@ def gen_variable_type_shard(out, aten_declarations, template_path, suffix, heade
                 wrapper_registrations.append(UNBOXEDONLY_WRAPPER_REGISTRATION.substitute(
                     declaration, formal_types=formal_types))
 
+            # TODO: ????? type mismatch in generated wrapper?
+            if declaration['name'] != 'polygamma':
+                # Emit ProfiledType code
+                profiled_body = emit_profiled_body(declaration)
+                profiled_method_definitions.append(METHOD_DEFINITION.substitute(
+                    declaration, type_definition_body=profiled_body))
+
+                if declaration['use_c10_dispatcher'] == 'full':
+                    profiled_wrapper_registrations.append(PROFILE_WRAPPER_REGISTRATION.substitute(
+                        declaration, formal_types=formal_types))
+                else:
+                    assert declaration['use_c10_dispatcher'] == 'unboxed_only'
+                    profiled_wrapper_registrations.append(PROFILE_UNBOXEDONLY_WRAPPER_REGISTRATION.substitute(
+                        declaration, formal_types=formal_types))
+
     env = {
         'type_derived_method_declarations': type_declarations,
         'type_derived_method_definitions': type_definitions,
         'wrapper_registrations': wrapper_registrations,
+        'profiled_method_definitions': profiled_method_definitions,
+        'profiled_wrapper_registrations': profiled_wrapper_registrations
     }
     if header:
         write(out, 'VariableType.h', VARIABLE_TYPE_H, env)
     else:
         write(out, 'VariableType%s.cpp' % suffix, VARIABLE_TYPE_CPP, env)
 
+    write(out, 'ProfiledType%s.cpp' % suffix, PROFILED_TYPE_CPP, env)
+
+def emit_profiled_body(declaration):
+    arguments = declaration['arguments']
+    returns = declaration['returns']
+    func = declaration['derivative']
+    name = declaration['name']
+    inplace = declaration['inplace']
+    is_out_fn = name.endswith('_out')
+    modifies_arguments = inplace or is_out_fn
+    returns_void = len(returns) == 0
+
+    processed_args = []
+    for a in arguments:
+        processed_args.append('{}'.format(a['name']))
+
+    if is_tensor_method(declaration):
```

Comment:
Not actionable comment on this PR: @smessmer, we should probably have some way of uniformly redispatching that doesn't require you to know if dispatch is going through a method or namespace.